planner/implementation/datasource.go

```diff
@@ -171,3 +171,90 @@ func NewIndexScanImpl(scan *plannercore.PhysicalIndexScan, tblColHists *statisti
 		tblColHists: tblColHists,
 	}
 }
+
+// IndexLookUpReaderImpl is the Implementation of PhysicalIndexLookUpReader.
+type IndexLookUpReaderImpl struct {
+	baseImpl
+
+	KeepOrder   bool
+	tblColHists *statistics.HistColl
+	extraProj   *plannercore.PhysicalProjection
+}
+
+// NewIndexLookUpReaderImpl creates a new table reader Implementation.
+func NewIndexLookUpReaderImpl(reader plannercore.PhysicalPlan, hists *statistics.HistColl, proj *plannercore.PhysicalProjection) *IndexLookUpReaderImpl {
+	base := baseImpl{plan: reader}
+	impl := &IndexLookUpReaderImpl{
+		baseImpl:    base,
+		tblColHists: hists,
+		extraProj:   proj,
+	}
+	return impl
+}
+
+// CalcCost calculates the cost of the table reader Implementation.
+func (impl *IndexLookUpReaderImpl) CalcCost(outCount float64, children ...memo.Implementation) float64 {
+	var reader *plannercore.PhysicalIndexLookUpReader
+	if impl.extraProj != nil {
+		impl.cost += impl.extraProj.GetCost(impl.extraProj.Stats().RowCount)
+	}
+	reader = impl.plan.(*plannercore.PhysicalIndexLookUpReader)
+	reader.IndexPlan, reader.TablePlan = children[0].GetPlan(), children[1].GetPlan()
+	reader.TablePlans = plannercore.FlattenPushDownPlan(reader.TablePlan)
+	reader.IndexPlans = plannercore.FlattenPushDownPlan(reader.IndexPlan)
```

Comment:
Put the `FlattenPushDownPlan` to `AttachChildren`?

Reply:
Good advice.